File: 2018/various/day02.hs
-- Advent of Code 2018 - Day 2
-- www.github.com/SondreSL

-- Part 1

count :: String -> Int
count (x:xs) = length (filter (x==) xs) + 1

findN :: Int -> String -> Int
findN _ "" = 0
findN n xs = if count xs == n
                then 1
                else findN n (tail xs)

solveA xs = res2 xs * res3 xs
    where
        res2 = sum . map (findN 2) . lines
        res3 = sum . map (findN 3) . lines

-- Not the correct answer. Off by 35, 8750 should be 8715.

-- Part 2
    


-- Main
main :: IO ()
main = do
    contents <- readFile "day02.txt"
    print $ solveA contents

File: 2018/Python/day02.py
# Advent of Code 2018 - Day 2
# www.github.com/SondreSL

# -- Part 1
from collections import Counter

with open('day02.txt') as f:
    twos = 0
    threes = 0
    for line in f:
        count = Counter(line).values()
        if 2 in count:
            twos += 1
        if 3 in count:
            threes += 1
    print('Part 1: ', twos * threes)


# -- Part 2
from itertools import combinations

with open('day02.txt') as f:
    lines = f.read().split()
    for a, b in combinations(lines, 2):
        eqs = [v != h for h, v in zip(a,b)]
        if sum(eqs) == 1:
            i = eqs.index(1)
            print('Part 2: ', a[:i] + a[i + 1:])
            break
